spec/abci++/abci++_basic_concepts_002_draft.md

```diff
@@ -1,51 +1,266 @@
+* [**VerifyVoteExtension:**](./abci++_methods_002_draft.md#verifyvoteextension) It allows validators to validate the vote extension data attached to a precommit message. If the validation fails, the precommit message will be deemed invalid and ignored
+by Tendermint. This has a negative impact on Tendermint's liveness, i.e., if repeatedly vote extensions by correct validators cannot be verified by correct validators, Tendermint may not be able to finalize a block even if sufficiently many (+2/3) of the validators send precommit votes for that block. Thus, `VerifyVoteExtension` should be used with special care.
+As a general rule, an Application that detects an invalid vote extension SHOULD
+accept it in `ResponseVerifyVoteExtension` and ignore it in its own logic. Tendermint calls it when 
+a process receives a precommit message with a vote extension.
 
-Additionally, there is a `Flush` method that is called on every connection,
-and an `Echo` method that is just for debugging.
+* [**FinalizeBlock:**](./abci++_methods_002_draft.md#finalizeblock) It delivers a decided block to the Application. The Application must execute the transactions in the block in order and update its state accordingly. Cryptographic commitments to the block and transaction results, via the corresponding
+parameters in `ResponseFinalizeBlock` are included in the header of the next block. Tendermint calls it when a new block is decided.
+
+### Mempool methods
+
+For validating new transactions before they are included in the mempool.
+
+* [**CheckTx:**](./abci++_methods_002_draft.md#checktx) This method allows the Application to validate transactions against its current state, e.g., checking signatures and account balances. If a transaction passes the validation, then tendermint adds it to its local mempool, discarding it otherwise. Tendermint calls it when it receives a new transaction either coming from an external user or another node.
+
+### Info methods
+
+For initialization and queries from the user.
+
+* [**Info:**](./abci++_methods_002_draft.md#info) Used to sync Tendermint with the Application during a handshake that happens on startup.
+
+* [**Query:**](./abci++_methods_002_draft.md#query) Clients can use this method to query the Application for information about the application state.
+
+### State-sync methods
+
+State sync allows new nodes to rapidly bootstrap by discovering, fetching, and applying
+state machine snapshots instead of replaying historical blocks. For more details, see the
+[state sync section](../p2p/messages/state-sync.md).
+
+New nodes will discover and request snapshots from other nodes in the P2P network.
+A Tendermint node that receives a request for snapshots from a peer will call
+`ListSnapshots` on its Application to retrieve any local state snapshots. After receiving
+ snapshots from peers, the new node will offer each snapshot received from a peer
```

Comment:
Are they snapshots or snapshot IDs/hashes?

Reply:
I would say none of the two but closer to snapshot IDs/hashes. As far as I understand, each entry of the returned list contains metadata about a snapshot: the height at which the snapshot was taken, some application-specific formatting data, the number of chunks, ...). Can someone confirm this? I can then fix the text :)

Reply:
I see. I would suggested to state it's snapshot metadata, and not the whole snapshot. My fear is that text implies the whole snapshot(s) is/are "offered" at this point

Reply:
I pushed a commit addressing this issue. I have a couple of questions:
- Assume a node receives a list of snapshots. What is the order in which these are offered to the application? Is this application-specific (if so what enables it), or there is some hard-coded order in Tendermint, e.g., height-decreasing order.
- Regarding safety, are snapshots signed by f+1 nodes or something like that? if this delegated to the application?

Reply:
> Regarding your second question, `OfferSnapshotRequest` contains field `app_hash` which is supposed to be the Merkle tree root of the state the snapshot represents. So the node can check it on the blockchain on one side (as a light client), and then, once it has reconstructed the state, it can re-calculate the hash to make sure the state is good.
> Someone else more familiar with state-sync may want to keep me honest here.

I think there is something I am missing here. If I understand your explanation, what that mechanism guarantees is that a node reaches the snapshot that was offered after applying all its chunks. What I do not understand is how a node knows that a given snapshot is legit. I am sure I am missing something fundamental. Maybe we can take this conversation out of the PR :)

My question I guess is